internal/sql/sql.go

```diff
@@ -0,0 +1,190 @@
+// args are for any placeholder parameters in the query.
+func (s *DB) Query(query string, args ...any) (*loggedRows, error) {
+	start := time.Now()
+	rows, err := s.db.Query(query, args...)
+	if dur := time.Since(start); dur > s.longQueryDuration {
+		s.log.Debug("slow query", zap.String("query", query), zap.Duration("elapsed", dur), zap.Stack("stack"))
+	}
+	return &loggedRows{rows, s.log.Named("rows"), s.longQueryDuration}, err
+}
+
+// queryRow executes a query that is expected to return at most one row.
+// QueryRow always returns a non-nil value. Errors are deferred until
+// Row's Scan method is called. If the query selects no rows, the *Row's
+// Scan will return ErrNoRows. Otherwise, the *Row's Scan scans the
+// first selected row and discards the rest.
+func (s *DB) QueryRow(query string, args ...any) *loggedRow {
+	start := time.Now()
+	row := s.db.QueryRow(query, args...)
+	if dur := time.Since(start); dur > s.longQueryDuration {
+		s.log.Debug("slow query row", zap.String("query", query), zap.Duration("elapsed", dur), zap.Stack("stack"))
+	}
+	return &loggedRow{row, s.log.Named("row"), s.longQueryDuration}
+}
+
+// transaction executes a function within a database transaction. If the
+// function returns an error, the transaction is rolled back. Otherwise, the
+// transaction is committed. If the transaction fails due to a busy error, it is
+// retried up to 15 times before returning.
+func (s *DB) Transaction(fn func(Tx) error) error {
+	var err error
+	txnID := hex.EncodeToString(frand.Bytes(4))
+	log := s.log.Named("transaction").With(zap.String("id", txnID))
+	start := time.Now()
+	attempt := 1
+	for ; attempt < maxRetryAttempts; attempt++ {
+		attemptStart := time.Now()
+		log := log.With(zap.Int("attempt", attempt))
+		err = s.transaction(s.db, log, fn)
+		if err == nil {
+			// no error, break out of the loop
+			return nil
+		}
+
+		// return immediately if the error is not a busy error
+		if !strings.Contains(err.Error(), s.dbLockedMsg) {
+			break
+		}
+		// exponential backoff
+		sleep := time.Duration(math.Pow(factor, float64(attempt))) * time.Millisecond
+		if sleep > maxBackoff {
+			sleep = maxBackoff
+		}
+		log.Debug("database locked", zap.Duration("elapsed", time.Since(attemptStart)), zap.Duration("totalElapsed", time.Since(start)), zap.Stack("stack"), zap.Duration("retry", sleep))
+		jitterSleep(sleep)
+	}
+	return fmt.Errorf("transaction failed (attempt %d): %w", attempt, err)
+}
+
+// Close closes the underlying database.
+func (s *DB) Close() error {
+	return s.db.Close()
+}
+
+// doTransaction is a helper function to execute a function within a transaction. If fn returns
+// an error, the transaction is rolled back. Otherwise, the transaction is
+// committed.
+func (s *DB) transaction(db *sql.DB, log *zap.Logger, fn func(tx Tx) error) error {
+	start := time.Now()
+	tx, err := db.Begin()
+	if err != nil {
+		return fmt.Errorf("failed to begin transaction: %w", err)
+	}
+	defer tx.Rollback()
```

Comment:
I was surprised as well when I saw it in `hostd` but it's a nice catch - all guarantee. `Rollback` will error if `Commit` was already called. I added some logging in case it fails.